torch/csrc/lazy/core/ir.cpp

```diff
@@ -52,28 +79,115 @@ Node::Node(OpKind op, size_t num_outputs, hash_t node_hash, std::function<hash_t
       node_hash_(node_hash),
       dag_hash_without_sizes_(dag_hash_fn(false)),
       dag_hash_with_sizes_(dag_hash_fn(true)),
-      metadata_(GetMetaDataIfDebugging()) {}
+      metadata_(GetMetaDataIfDebugging()),
+      python_stacktrace_(GetFirstUserFrameInPythonIfEnabled()) {}
 
 Node::Node(OpKind op, size_t num_outputs, std::function<hash_t(bool)> node_hash_fn)
     : op_(op),
       num_outputs_(num_outputs),
       node_hash_(node_hash_fn(!enableDynamicShape())),
       dag_hash_without_sizes_(node_hash_fn(false)),
       dag_hash_with_sizes_(node_hash_fn(true)),
-      metadata_(GetMetaDataIfDebugging()) {}
+      metadata_(GetMetaDataIfDebugging()),
+      python_stacktrace_(GetFirstUserFrameInPythonIfEnabled()) {}
+
+
+Node::Node(OpKind op, OpList operands, std::vector<Shape>&& shapes,
+           size_t num_outputs, hash_t hash_seed)
+    : Node(op, num_outputs,
+           // TODO(WHC) this is inefficient (having to compute node_hash twice
+           // since I can't call hash() yet) so probably move dag_hash
+           // initialization to a separate function?
+           /* node_hash */ HashCombine(op.hash(), hash_seed),
+           /* dag_hash */
+           [&](bool bakeInSizes) { return OperandHashes(operands, HashCombine(op.hash(), hash_seed), bakeInSizes); }
```

Comment:
I am very against the idea of exposing more `hash_with_shape` and `hash_without_size` concept to the base node class, because I still think it is ts implementation detail and does not apply to xla.

@shunting314 I am OK with for the sake of not blocking this pr for too long. Is the plan to remove this soon?

Reply:
I'm not sure if @shunting314 will be able to get to it soon, I might help instead.  But I am not sure if it makes sense to move shape cache to Node base.  We may want to do it, but it would be good to do it separately from this PR and discuss more with torch/xla first about how they would use it.

@antoniojkim do you think it would be possible to make this PR a little smaller by keeping the shape cache in TSNode?

Reply:
Before you make any changes, I think it's a good idea to chat with @JackCaoG and see if we are going to use the shape cache for XLA.  if so then probably makes sense to land this as-is and then clean up the hash_with_sizes stuff.